connect-location/src/main/java/com/ifttt/location/AwarenessGeofenceProvider.java

```diff
@@ -40,105 +51,145 @@
+                new DiffCallback() {
+                    @Override
+                    public void onAddFence(String key, AwarenessFence value, PendingIntent pendingIntent) {
+                        requestBuilder.addFence(key, value, pendingIntent);
+                    }
+
+                    @Override
+                    public void onRemoveFence(String key) {
+                        requestBuilder.removeFence(key);
+                    }
+                }
+            );
+            fenceClient.updateFences(requestBuilder.build());
+        });
+    }
+
+    private static String getEnterFenceKey(String id) {
+        return id.concat("/enter");
+    }
 
-        List<UserFeatureField> userFeatureFields = new ArrayList<>();
+    private static String getExitFenceKey(String id) {
+        return id.concat("/exit");
+    }
 
-        for (Feature feature : connection.features) {
-            if (feature.userFeatures == null) {
+    interface DiffCallback {
+        void onAddFence(String key, AwarenessFence value, PendingIntent pendingIntent);
+
+        void onRemoveFence(String key);
+    }
+
+    @VisibleForTesting
+    static void diffFences(
+        Connection.Status state,
+        List<Feature> features,
+        Set<String> registeredFenceKeys,
+        PendingIntent enterPendingIntent,
+        PendingIntent exitPendingIntent,
+        DiffCallback callback
+    ) {
+        Set<String> fenceKeysToRemove = new HashSet<>(registeredFenceKeys);
+
+        for (Feature feature : features) {
+            if (feature.userFeatures == null || state != Connection.Status.enabled) {
+                // Skip registration if the UserFeature is null or the connection is not enabled.
                 continue;
             }
 
             for (UserFeature userFeature : feature.userFeatures) {
+                if (!userFeature.enabled) {
+                    // Skip processing disabled UserFeatureSteps.
+                    continue;
+                }
+
                 for (UserFeatureStep step : userFeature.userFeatureSteps) {
                     for (UserFeatureField userFeatureField : step.fields) {
                         if (!locationFieldTypesList.contains(userFeatureField.fieldType)) {
                             continue;
                         }
 
-                        userFeatureFields.add(userFeatureField);
-                        if (userFeatureField.fieldType.equals(FIELD_TYPE_LOCATION_ENTER_EXIT)) {
-                            updatedIds.add(getEnterFenceKey(userFeatureField.fieldId));
-                            updatedIds.add(getExitFenceKey(userFeatureField.fieldId));
-                        } else {
-                            updatedIds.add(userFeatureField.fieldId);
+                        if (!(userFeatureField.value instanceof LocationFieldValue)) {
+                            throw new IllegalStateException("Location field should have LocationFieldValue data.");
+                        }
```

Comment:
is this necessary? It feels like an additional check if we already know that the field type is one of the geofence types.

Reply:
Yeah you're right, I don't think we really need this here. If there is some errors on the data side, the parser will likely catch it and crash.